### DISCUSSION TITLE
Keep only one identifier in PID.3 repeating field

### DISCUSSION THREAD

**Discussion by amaster507** [nextgenhealthcare/connect#5450] [Q&A] [answered]
I had a `PID` segment that has multiple repetition for `PID.3`

```hl7
PID|1||Mxxxxxxxxx^^^LAB&2.16.840.1.x&ISO^MR^ABCD&YYYYYYYYYY&CLIA~nnn-nn-nnnn^^^LAB&2.16.840.1.x&ISO^SS^ABCD&YYYYYYYYYY&CLIA~Sxxxxxxxx^^^LAB&2.16.840.1.x&ISO^XX^ABCD&YYYYYYYYYY&CLIA~HVxxxxxxxxxxx^^^LAB&2.16.840.1.x&ISO^AN^ABCD&YYYYYYYYYY&CLIA||TEST^TEST^^^^^L|
```

I am needing to keep only one of these subcomponents as identified by `PID.3.5`

At first I tried to copy the fields into variables and then write these variables to the main PID.3 field. But this was not working due to some fields being basic string while others were xml structured deeper.

For posterity sake, I am sharing my completed solution so that others (and I next time) can benefit when needing this seemingly common and not too easy task.

```js
var pid3_found = undefined;
var identifier_needed = "MR"
for each (pid3 in msg['PID']['PID.3']) {
  if (pid3['PID.3.5'].toString()===identifier_needed ) {
    pid3_found = pid3.copy()
  }
}
if (pid3_found !== undefined) {
  msg['PID']['PID.3'] = pid3_found
} else {
  // throw error or do other alternative action
}
```

results:

```hl7
PID|1||Mxxxxxxxxx^^^LAB&2.16.840.1.x&ISO^MR^ABCD&YYYYYYYYYY&CLIA||TEST^TEST^^^^^L|
```

I believe this is safe for my use case, because my source always includes repeated PID.3 fields. If your HL7 source does not always generate PID, PID.3, or repetitions in PID.3 then you may need to add in some additional error handling.

Hope this helps someone else (or even myself) in the future.

**Comment by tonygermano** [comment, marked as answer]
FYI, I also created a code template to handle stuff like this. You can find it here https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Filter%20XMLLists

In your case, you could do this
```js
var identifier_needed = "MR"
var pid3mr = xFilter(msg['PID']['PID.3'], pid3 => pid3['PID.3.5'].toString() === identifier_needed)
if (pid3mr.length() === 1) {
  msg['PID']['PID.3'] = pid3mr
}
else {
  // this could be 0 or more than 1. Your code would return only the last one found if there were more than 1
}
```

If you use some of the additional parameters and string coercion, it can be shortened to
```js
// This will automatically throw a RangeError if exactly one match is not found.
msg['PID']['PID.3'] = xFilter(msg['PID']['PID.3'], pid3 => pid3['PID.3.5'] == 'MR', 1, 1, true)

// Removing the 'true' at the end would only throw an error if 0 were found. It would select
// only the first repetition that matched if there were multiples.
```

**Reply by amaster507** [reply]
I went looking through the example code templates, but must have overlooked this one. I wasn't thinking of the problem as a filtering problem, but rather a copy and replace problem. Thank you for the insight.
